### ISSUE TITLE
[Patterns][records][inference] ECJ never computes the upwards projection for var type inference in record patterns

### ISSUE THREAD

**Issue by srikanth-sankaran** [eclipse-jdt/eclipse.jdt.core#2007]
This is a follow up from https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2004

JLS 14.30.1 reads:

```
The type of a pattern variable declared in a nested type pattern is determined as
follows:
• If the LocalVariableType is UnannType then the type of the pattern variable is
denoted by UnannType.
• If the LocalVariableType is var then the type pattern must appear directly in the
pattern list of a record pattern, or a compile-time error occurs.
Let R be the type of the record pattern, and let T be the type of the corresponding
component field in R (§8.10.3). The type of the pattern variable is the upward
projection of T with respect to all synthetic type variables mentioned by T.
```

If however you follow the steps in https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2005#issuecomment-1938213236
you will see that the exception is never thrown - essentially the line with the call
`this.local.type.resolvedType = recType.upwardsProjection(scope, mentionedTypeVariables);` is never reached.

This was behavior as of **4.30** also of master as of date.

A few points to investigate:

1. `org.eclipse.jdt.internal.compiler.ast.RecordPattern.resolveType(BlockScope)` reads:

```
              if (this.resolvedType.isRawType()) {
			TypeBinding expressionType = expectedType();
			if (expressionType instanceof ReferenceBinding) {
				ReferenceBinding binding = inferRecordParameterization(scope, (ReferenceBinding) expressionType);
				if (binding == null || !binding.isValidBinding()) {
					scope.problemReporter().cannotInferRecordPatternTypes(this);
				    return this.resolvedType = null;
				}
				this.resolvedType = binding;
			}
		}
```

I suspect the last line should actually be `this.resolvedType = binding.capture(scope, this.sourceStart, this.sourceEnd);`

2. I think the capture call down below in the same method is misplaced.
3. There _could_ also be something wrong with `org.eclipse.jdt.internal.compiler.lookup.ReferenceBinding.upwardsProjection(Scope, TypeBinding[])`  - I found some strange results in experiments but didn't capture (!) (!!) them :)

**Comment by srikanth-sankaran** [comment]
@stephan-herrmann, @mpalat - FYI. I may need consulting help later when I take this up.

**Comment by srikanth-sankaran** [comment]
Any fix for this must start by unrolling the stop gap fix put in by https://github.com/srikanth-sankaran/eclipse.jdt.core/commit/a79934dbbdd949bb6373df3827313696fad8bb5b - this fix is really restoring 4.30 behavior rather than spec compliant behavior.

**Comment by srikanth-sankaran** [comment]
> This is a follow up from #2004

[...]

> 3. There _could_ also be something wrong with `org.eclipse.jdt.internal.compiler.lookup.ReferenceBinding.upwardsProjection(Scope, TypeBinding[])`  - I found some strange results in experiments but didn't capture (!) (!!) them :)

I found out what is wrong with the upwardsProjection call. We should be calling upwardsProjection on the **component* type and not the _record_ type!
